tests/test-backend-ops.cpp

```diff
@@ -317,6 +323,339 @@ enum test_mode {
+                const std::string& test_mode, bool supported, bool passed, const std::string& error_message = "",
+                double time_us = 0.0, double flops = 0.0, double bandwidth_gb_s = 0.0,
+                size_t memory_kb = 0, int n_runs = 0)
+        : backend_name(backend_name), op_name(op_name), op_params(op_params), test_mode(test_mode),
+          supported(supported), passed(passed), error_message(error_message), time_us(time_us),
+          flops(flops), bandwidth_gb_s(bandwidth_gb_s), memory_kb(memory_kb), n_runs(n_runs) {
+        // Set test time
+        time_t t = time(NULL);
+        char buf[32];
+        std::strftime(buf, sizeof(buf), "%FT%TZ", gmtime(&t));
+        test_time = buf;
+
+        // Set build info
+        build_commit = LLAMA_COMMIT;
+        build_number = LLAMA_BUILD_NUMBER;
+    }
+
+    static const std::vector<std::string> & get_fields() {
+        static const std::vector<std::string> fields = {
+            "test_time", "build_commit", "build_number", "backend_name", "op_name", "op_params", "test_mode",
+            "supported", "passed", "error_message", "time_us", "flops",
+            "bandwidth_gb_s", "memory_kb", "n_runs"
+        };
+        return fields;
+    }
+
+    enum field_type { STRING, BOOL, INT, FLOAT };
+
+    static field_type get_field_type(const std::string & field) {
+        if (field == "supported" || field == "passed") {
+            return BOOL;
+        }
+        if (field == "memory_kb" || field == "n_runs" || field == "build_number") {
+            return INT;
+        }
+        if (field == "time_us" || field == "flops" || field == "bandwidth_gb_s") {
+            return FLOAT;
+        }
+        return STRING;
+    }
+
+    std::vector<std::string> get_values() const {
+        return {
+            test_time,
+            build_commit,
+            std::to_string(build_number),
+            backend_name,
+            op_name,
+            op_params,
+            test_mode,
+            std::to_string(supported),
+            std::to_string(passed),
+            error_message,
+            std::to_string(time_us),
+            std::to_string(flops),
+            std::to_string(bandwidth_gb_s),
+            std::to_string(memory_kb),
+            std::to_string(n_runs)
+        };
+    }
+};
+
+// Printer classes for different output formats
+struct printer {
+    virtual ~printer() {}
+    FILE * fout = stdout;
+    virtual void print_header() {}
+    virtual void print_test_result(const test_result & result) = 0;
+    virtual void print_footer() {}
+
+    // General purpose output methods
+    virtual void print_info(const char * format, ...) = 0;
+    virtual void print_error(const char * format, ...) = 0;
+    virtual void print_device_info(const char * format, ...) = 0;
+    virtual void print_test_summary(const char * format, ...) = 0;
+    virtual void print_status_ok() = 0;
+    virtual void print_status_fail() = 0;
```

Comment:
I don't think this a good design, all of these functions do the same could be replaced with a single `print_message`. If you want transfer the responsibility of formatting the output to a class, then the class needs to have the information to decide what to print, not an unstructured message. Ideally all of the functions would receive an object to print, and the class would determine how to format that object.

Reply:
Sounds good. I’ll update the code accordingly.

Reply:
No, this way still depends on the llama.cpp common lib.

Reply:
Please check if I understand this correctly: https://github.com/ggml-org/llama.cpp/pull/14368/commits/f4f551222cb2edcef8ceaa6b711640f325ba1bdd
Thanks.